rfd/0030-client-side-session-encryption.md

```diff
@@ -0,0 +1,168 @@
+
+Example YAML:
+
+```yaml
+teleport:
+  storage:
+    audit_sessions_uri: s3://teleport/recordings?region=ap-southeast-2
+    audit_sessions_local_keys:
+      - some-key:main:AgqHAQf+tym2eV9iSKWS5TMvcaKYFFX5F3DQxICWvKk=
+      - another-key:GhBT5ktJiXGaxnFyAmzRkt0K
+      - key-in-a-file:/path/to/some/key
+```
+If a `audit_sessions_local_keys` entry is not present, the system will fall back to the existing gzipped protobuf format.
+
+If the `audit_sessions_local_keys` is malformed in any way, including any key files it references, Teleport will fail to start and alert the user with an error message.
+
+#### Encryption & Decryption
+
+As per [RFD-002](./0002-streaming.md), the recording system gathers session events and compresses them into _slices_. Once the compressed slice reaches a minimum size (currently 5 MiB) the chunk is wrapped with a slice header and stored, before moving on to the next slice.  This finalization step provides a natural place to encrypt the slice. When operating in encrypted mode, the slice writer will 
+ 1. Envelope encrypt the gzipped event records using the active encryption key
+ 2. Rewrite the slice record to treat the resulting ciphertext as the slice payload, and 
+ 3. Pass it though to the existing storage machinery
+
+The same process happens in reverse. When the recording reader encounters a new slice record, it will
+ 1. Examine the slice to see if the slice is encrypted
+ 2. If so, extract the Key-Encryption Key ID from the envelope header (see below)
+ 3. Map the key ID to a key supplied in the config file, and use that key to decrypt the envelope
+ 4. Pass the resulting plaintext (i.e. a gzipped series of protobuf records) along as per the existing recording system.
+
+Each slice may be encrypted with a different master key, even in the same file. This allows multiple nodes with different encryption keys to write to the same session. The session will be played back as long as the node managing the playback has access to all the keys used for the session.
+
+Reusing the existing recording machinery also allows session files in transit to be recorded at rest whenever they are stored locally on a node (e.g. async recording).
+
+#### V2 Proto Recording Format
+
+[RFD-002](./0002-streaming.md), shows that a recording file is a sequence of semi-independent _slices_. Each slice contains a gzipped sequence of session events serialised into the protobuf wire format. Each slice is at least (and probably close to) 5MiB for all but the _last_ slice in a recording, which may be any size. No upper bound on slice size is currently enforced.
+
+The V2 format will add a `flags` field after the version header. For consistency's sake, this will be another Uint64. The new header will look this:
+
+| Field Size (bytes) | Encoding |         Interpretation              |
```

Comment:
specifically wondering, what would be required to have this supported out of the box:

https://docs.aws.amazon.com/athena/latest/ug/parquet-serde.html

and 

https://docs.aws.amazon.com/athena/latest/ug/compression-formats.html

Reply:
@xacrimon here is what comes to mind:

* Without encryption, if we find a way to present each independent chunk as a part of the larger gzip/bzip or SNAPPY, then the actual chunk to be in parquet format, then anyone unpacking the resulting .snappy S3 file will see a very simple archive file with parquet inside.

This will require some wizardry during creation of chunks, e.g. creating a special padding parquet event at the end perhaps or similar. I think it's worth exploring. Now, the client side encryption can easily break this compatibility, if it creates its own, independent envelope with encryption, after the data has been gzipped. 

I think it's worth discussing at this stage with @tcsc and may be prototyping together, so we end up with something that we won't have to rewrite again. cc @russjones.

Reply:
For example, snappy supports frames, designed for streaming over HTTP chunks:

https://github.com/google/snappy/blob/master/framing_format.txt

If we represent each slice as a snappy frame, and the internal contents as parquet, the resulting concatenated S3 file, could be a proper snappy parquet file.

Then the client side encryption will encrypt snappy frames using go's envelopes and may be snappy frames again with no additional compression? Then the resulting encrypted file will be a snappy archive of encrypted snappy frames :)

Something like that could work, what do you think?

Reply:
That seems doable. Parquet SerDe doesn't look compatible with encrypted files however. Compressing snappy frames looks really neat and it looks like one of the better options.